src/EditorFeatures/Core/Implementation/Diagnostics/AbstractDiagnosticsTaggerProvider.AggregatingTagger.cs

```diff
@@ -174,10 +174,6 @@ private void OnDiagnosticsUpdated(object sender, DiagnosticsUpdatedArgs e)
 
             private void OnDiagnosticsUpdated(DiagnosticsUpdatedArgs e)
             {
-                // First see if this is a document/project removal.  If so, clear out any state we
-                // have associated with any analyzers we have for that document/project.
-                ProcessRemovedDiagnostics(e);
```

Comment:
I don't think the problem here is the diagnostic service or the diagnostic update events. The events are pretty self contained, with information about document/project/workspace to which they belong.

I think the _real_ issue here is that we are trying to share the same [AggregatingTagger](http://source.roslyn.io/Microsoft.CodeAnalysis.EditorFeatures/R/564374b6f28b6ec7.html) instance across workspaces. I think this can cause all sorts of race conditions based on the order in which workspace events and/or diagnostic updated events are generated when a document moves across workspace (which is very common the way Miscellaneous workspace works).

Current design of having a single AggregatingTagger seems flawed. We need either of the following designs:

1. Have a single AggregatingTagger instance per subject buffer, per workspace OR
2. AggregatingTagger should have its state be separated per-workspace. For example, the key for [_idToProviderAndTagger](http://source.roslyn.io/Microsoft.CodeAnalysis.EditorFeatures/R/6ef88cdf69bdbd8d.html) needs to include the workspace and [TagsChanged](http://source.roslyn.io/Microsoft.CodeAnalysis.EditorFeatures/R/799dac593da0b61f.html) event args should include information about the workspace in which the tags changed for the subject buffer.